# Recommender System

1. Hitung similaritas rating Alesya dengan user lainnya menggunakan Pearson Correlation dan pilih 1 lagi metode hitung similaritas dari metode berikut:
a. Cosine Similarities
b. Euclidean Distance
c. Manhattan Distance
2. Berikan penjelasan/representasi dari nilai similaritas yang diperoleh.
3. Gunakan hasil similaritas dari Pearson Correlation untuk menghitung prediksi rating yang akan diberikan Alesya terhadap item Comedy3, dengan menggunakan threshold 0,5.
4. Beri penjelasan/representasi dari hasil prediksi rating Alesya terhadap item Comedy3.

In [1]:
import pandas as pd
import numpy as np
import scipy
from scipy import stats
from scipy.spatial import distance

# Pearson Correlation

In [2]:
data = pd.read_excel('RecSys.xlsx')
data

User  Film1  Film2  Film3  Drama1  Drama2  Drama3  Comedy1  Comedy2  \
0   Alesya    3.0    4.0    2.0     1.0     2.0     1.0      3.0      5.0   
1    User1    2.0    5.0    1.0     1.0     3.0     4.0      4.0      4.0   
2    User2    5.0    4.0    2.0     4.0     1.0     1.0      5.0      1.0   
3    User3    3.0    4.0    2.0     2.0     1.0     2.0      5.0      1.0   
4    User4    1.0    3.0    1.0     5.0     2.0     4.0      4.0      2.0   
5    User5    2.0    2.0    1.0     3.0     3.0     1.0      3.0      3.0   
6    User6    5.0    1.0    1.0     2.0     4.0     1.0      2.0      4.0   
7    User7    2.0    1.0    1.0     1.0     1.0     4.0      1.0      1.0   
8    User8    3.0    1.0    4.0     1.0     4.0     2.0      1.0      1.0   
9    User9    4.0    5.0    2.0     4.0     5.0     1.0      2.0      5.0   
10  User10    1.0    3.0    5.0     3.0     1.0     2.0      2.0      5.0   
11  User11    2.0    2.0    5.0     2.0     4.0     3.0      4.0      3.0   
12  User12    4.0    2.0    1.0     1.0     2.0     3.0      5.0      5.0   
13  User13    5.0    5.0    3.0     4.0     3.0     5.0      1.0      3.0   
14  User14    5.0    4.0    2.0     5.0     1.0     3.0      4.0      1.0   
15  User15    2.0    1.0    5.0     2.0     5.0     2.0      5.0      4.0   
16  User16    3.0    3.0    2.0     3.0     1.0     1.0      3.0      2.0   
17  User17    4.0    4.0    3.0     4.0     4.0     4.0      1.0      5.0   
18  User18    4.0    1.0    1.0     2.0     2.0     2.0      2.0      4.0   
19  User19    5.0    2.0    4.0     1.0     4.0     1.0      4.0      3.0   
20  User20    1.0    2.0    3.0     1.0     3.0     1.0      3.0      2.0   
21  User21    4.0    3.0    3.0     3.0     2.0     4.0      3.0      5.0   
22  User22    3.0    4.0    1.0     4.0     1.0     5.0      4.0      4.0   
23  User23    2.0    1.0    5.0     2.0     5.0     4.0      1.0      5.0   
24  User24    5.0    4.0    1.0     3.0     1.0     5.0      5.0      2.0   
25  User25    2.0    2.0    5.0     4.0     5.0     1.0      2.0      1.0   
26  User26    3.0    2.0    1.0     5.0     2.0     2.0      5.0      4.0   
27  User27    5.0    4.0    3.0     5.0     5.0     3.0      2.0      2.0   
28  User28    2.0    3.0    4.0     5.0     1.0     4.0      5.0      4.0   
29  User29    1.0    3.0    3.0     3.0     2.0     1.0      2.0      5.0   
30  User30    4.0    1.0    3.0     3.0     5.0     5.0      5.0      3.0   

    Comedy3  
0       NaN  
1       2.0  
2       3.0  
3       4.0  
4       1.0  
5       1.0  
6       5.0  
7       1.0  
8       5.0  
9       3.0  
10      5.0  
11      2.0  
12      3.0  
13      5.0  
14      3.0  
15      5.0  
16      1.0  
17      2.0  
18      3.0  
19      4.0  
20      2.0  
21      4.0  
22      5.0  
23      3.0  
24      2.0  
25      5.0  
26      2.0  
27      3.0  
28      5.0  
29      2.0  
30      4.0

In [3]:
hitungPearson=[]
hitungPearson = data
hitungPearson['sim_pearson'] = 1

x = np.array(data.iloc[0,1:9])

for i in range (1,len(hitungPearson)):
  y = np.array(data.iloc[i,1:9])
  sim = scipy.stats.pearsonr(x, y)
  sim = sim[0]
  hitungPearson.iloc[i:,10]=sim

In [4]:
#menghilangkan baris dengan nilai similarity dibawah 0.5
hitungPearson.drop(hitungPearson[hitungPearson.sim_pearson <0.5].index, inplace=True)
hitungPearson

User  Film1  Film2  Film3  Drama1  Drama2  Drama3  Comedy1  Comedy2  \
0   Alesya    3.0    4.0    2.0     1.0     2.0     1.0      3.0      5.0   
1    User1    2.0    5.0    1.0     1.0     3.0     4.0      4.0      4.0   
9    User9    4.0    5.0    2.0     4.0     5.0     1.0      2.0      5.0   
12  User12    4.0    2.0    1.0     1.0     2.0     3.0      5.0      5.0   
29  User29    1.0    3.0    3.0     3.0     2.0     1.0      2.0      5.0   

    Comedy3  sim_pearson  
0       NaN     1.000000  
1       2.0     0.536925  
9       3.0     0.537856  
12      3.0     0.594759  
29      2.0     0.581238

In [5]:
#hasil prediksi rating alesya berdasarkan pearson correlation
a = len(hitungPearson)
pred_rat_alesya_com3 = sum(hitungPearson.iloc[1:a,9]) / (a-1)
pred_rat_alesya_com3

2.5

# Euclidean Similarity

In [6]:
data = pd.read_excel('RecSys.xlsx')

hitungEu=[]
hitungEu = data
hitungEu['sim_cos'] = 1

x = np.array(data.iloc[0,1:9])

for i in range (1,len(hitungEu)):
  y = np.array(data.iloc[i,1:9])
  sim = np.linalg.norm(x-y)
  hitungEu.iloc[i:,10]=sim

In [7]:
#menghilangkan baris dengan nilai similarity diatas 4
hitungEu.drop(hitungEu[hitungEu.sim_cos > 4].index, inplace=True)
hitungEu

User  Film1  Film2  Film3  Drama1  Drama2  Drama3  Comedy1  Comedy2  \
0   Alesya    3.0    4.0    2.0     1.0     2.0     1.0      3.0      5.0   
1    User1    2.0    5.0    1.0     1.0     3.0     4.0      4.0      4.0   
5    User5    2.0    2.0    1.0     3.0     3.0     1.0      3.0      3.0   
12  User12    4.0    2.0    1.0     1.0     2.0     3.0      5.0      5.0   
16  User16    3.0    3.0    2.0     3.0     1.0     1.0      3.0      2.0   
18  User18    4.0    1.0    1.0     2.0     2.0     2.0      2.0      4.0   
21  User21    4.0    3.0    3.0     3.0     2.0     4.0      3.0      5.0   
29  User29    1.0    3.0    3.0     3.0     2.0     1.0      2.0      5.0   

    Comedy3   sim_cos  
0       NaN  1.000000  
1       2.0  3.872983  
5       1.0  3.872983  
12      3.0  3.741657  
16      1.0  3.872983  
18      3.0  3.872983  
21      4.0  4.000000  
29      2.0  3.316625

In [8]:
#prediksi rating alesya berdasarkan euclidian
a = len(hitungEu)
pred_cos_rat_alesya_com3 = sum(hitungEu.iloc[1:a,9]) / (a-1)
pred_cos_rat_alesya_com3

2.2857142857142856

# Resume

1. Similarity pertama menggunakan metode Pearson Correlation antara 2 data menghasilkan correlation diantara 0 sampai 1, sehingga pada saat diberi threshold 0.5, maka data dengan nilai similarity diatas 0.5 dapat dikatakan mirip. Kemudian nilai tersebut dievaluasi menggunakan average.

2. Kemudian untuk similarity yang kedua dengan menggunakan metode Euclidian Distance, menggunakan 4 sebagai threshold, sehingga baris data dengan similarity diatas 4 akan di evaluasi.

3. Dengan melakukan rata-rata dari baris data yang terpilih dengan metode pearson correlation dan euclidien distance, didapat nilai rating 2.5 dan 2.29. Sehingga dapat diambil kesimpulan bahwa Comedy3 tidak terlalu recommended.